docs/mmds/mmds-user-guide.md

```diff
@@ -0,0 +1,240 @@
+            "latest": {
+                  "meta-data": {
+                       "ami-id": "ami-87654321",
+                       "reservation-id": "r-79054aef",
+                  }
+            }      
+    }"
+```
+
+# Retrieving metadata
+
+MicroVM metadata can be retrieved:
+* from host operating system, communicating with the firecracker API server.
+* from guest operating system, communicating directly with MMDS.
+
+For the scope of this chapter, let's assume the data store content is the JSON
+below:
+
+```json
+{
+    "latest": {
+          "meta-data": {
+               "ami-id": "ami-87654321",
+               "reservation-id": "r-79054aef"
+          }
+    }
+}  
+```
+
+## Retrieving metadata in the host operating system
+
+To retrieve existing metadata in host operating system, a HTTP `GET` request to
+the `/mmds` resource must be issued. The HTTP response returns the existing
+metadata, structured in JSON format. A complete description of
+retrieving metadata firecracker API can be found in the
+[firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+### Example
+
+```bash
+curl -s --unix-socket /tmp/firecracker.socket http://localhost/mmds
+```
+
+Output:
+
+```json
+{
+    "latest": {
+          "meta-data": {
+               "ami-id": "ami-87654321",
+               "reservation-id": "r-79054aef"
+          }
+    }
+}  
+```
+
+## Retrieving metadata in the guest operating system
+
+To retrieve existing metadata in guest operating system, a HTTP `GET` request
+must be issued. The requested resource can be referenced by its corresponding
+[JSON Pointer](https://tools.ietf.org/html/rfc6901), which is also the path of
+the MMDS request. The HTTP response content will contain the referenced metadata
+resource, as a JSON formatted text.
+
+### Example
+
+Retrieving the `meta-data` resource:
+```bash
+MMDS_IPV4_ADDR=169.254.170.2
+RESOURCE_POINTER=latest/meta-data
+curl -s http://${MMDS_IPV4_ADDR}/${RESOURCE_POINTER}
+```
+
+Output: 
+
+```json
+{
+    "ami-id": "ami-87654321",
+    "reservation-id": "r-79054aef"
+}
+```
+
+# Appendix
+
+#### Example use case: credential rotation
+
+For this example, the guest expects to find some sort of credentials (say, a
+secret access key) by issuing a `GET` request to
+`http://169.254.169.254/latest/meta-data/credentials/secret-key`. Most similar
+use cases will encompass the following sequence of steps:
+
+1. Some agent running on the host sends a `PUT` request with the initial
+   contents of the MMDS, using the Firecracker API. This most likely takes
+   place before the microVM starts running, but may also happen at a later
+   time. Guest MMDS requests which arrive prior to contents being available
+   receive a *NotFound* response.
+1. The contents are saved to the data store.
+1. The guest sends a `GET` request for the secret key, which is intercepted by
+   the device model.
+1. The device model queries the data store, and gets the value associated with
```

Comment:
device model queries the datastore using dumbo stack
A HTTP response is assembled and sent back to the guest, using dumbo stack and device model.

Reply:
I agree, I think this description is too implementation specific. We can also simplify and not mention any of the MMDS internal hooks (device model, dumbo stack etc). From the user perspective, I think it is important what is the flow of this use case (credential rotation), and what input the user needs to provide to make it happen. I would suggest the following rephrasing:

"MMDS receives the `GET` request and in return, it assembles a HTTP response and sends it back to the guest." 

We can simplify further by explaining what we mean by the guest, replacing `guest` with `guest application` or `MMDS client`, to make it more clear where the `GET` request originates from.

What do you think?

Reply:
If we don't specify it here, we should specify it in design so that the dataflow would be clear to all users.